### ISSUE TITLE
Code coverage is incorrect

### ISSUE THREAD

**Issue by lyyourc** [vire/jest-vue-preprocessor#15]
expected `100%` , but only got `93.75%` ：

<img width="867" alt="screen shot 2017-04-01 at 12 45 28 pm" src="https://cloud.githubusercontent.com/assets/6823863/24575478/447c2ce4-16d9-11e7-981d-1d9cb195accc.png">

## P.S.

* jest: `^19.0.2`
* jest-vue-preprocessor: `^0.2.0`,
* Node: `7.8`

```js
// Hello.spec.js
import Vue from 'vue'
import Hello from '@/components/Hello'

describe('Hello.vue', () => {
  it('should render correct contents', () => {
    const Constructor = Vue.extend(Hello)
    const vm = new Constructor().$mount()
    expect(vm.$el.querySelector('.hello h1').textContent)
      .toBe('Welcome to Your Vue.js App')
  })
})
```

```js
// package.json
"jest": {
    "collectCoverage": true,
    "moduleNameMapper": {
      "src/components/([^\\.]*)$": "<rootDir>/src/components/$1.vue",
      "^vue$": "vue/dist/vue.common.js",
      "src/([^\\.]*)$": "<rootDir>/src/$1.vue",
      "(.*)/(.*)$": "$1/$2.vue"
    },
    "moduleFileExtensions": [
      "js",
      "vue"
    ],
    "transform": {
      "^.+\\.js$": "<rootDir>/node_modules/babel-jest",
      ".*\\.(vue)$": "<rootDir>/node_modules/jest-vue-preprocessor"
    }
  },
```

**Comment by mathieumure** [comment]
Hi,
I've spent a lot of time on the wrong coverage values and on the spawn of the undefined file on my root folder ; and I've finally why the coverage doesn't work.

## TL;DR
Remove the plugin `istanbul` from your `.babelrc` file

## The cause

When you generate a project from the vue-webpack template, the babel configuration automatically includes the istanbul plugin for the test env. This plugin is mandatory when your want the coverage with mocha and chai.
However, Jest comes with istanbul and will add hook functions (needed by istanbul) on the transpiled code produced by Babel. Babel is also configured to add hook functions due to the istanbul plugin.
That's why the coverage computed by jest's istanbul is not your actual code but your code + the babel's istanbul code!
As the istanbul babel plugin is not needed by jest you can remove it from the .babelrc file.

Hope it will save you, hours of search.

**Comment by johnsardine** [comment]
@lyyourc I also have this problem. I was able to get it working, kind of, but then the coverage does not match 100%. 

I have created a fork https://github.com/johnsardine/jest-vue-preprocessor that provides a source map to the jest transform

I also created an installation with that package https://github.com/johnsardine/vue-jest

Can you try it and give some feedback?

Perhaps the author of this repo already has an idea of how to fix this

**Comment by vire** [comment]
@johnsardine I started looking at the sourcemaps issue today, so as I got it right, you need as output 2 files (src+sorucemaps) to get code coverage working correctly or, the .babelrc `"sourceMaps": "inline"` does not work with the coverage?

**Comment by johnsardine** [comment]
@vire With `sourceMaps` set to inline, the result is the same as if you had no map, although [Jest mapCoverage](https://facebook.github.io/jest/docs/configuration.html#mapcoverage-boolean) says it accepts inline sourcemaps as well, I got better results with the map separated. Let me know if you need more clarification.

**Comment by eddyerburgh** [comment]
Hi @vire. I think this preprocessor needs correct sourcemaps for .vue files. This is different to babel source maps.

To get source map with .vue files you need to generate a source map inside vue jest preprocessor.

You can see logic in vueify for reference - https://github.com/vuejs/vueify/blob/master/lib/compiler.js#L187

**Comment by vire** [comment]
@eddyerburgh I'll try to dig into in the next couple of days.

**Comment by Xiphe** [comment]
That would be awesome! Let me know when you need any help or rather would like a PR instead of own digging :)

**Comment by vire** [comment]
@Xiphe every PR is welcome here, when you have time you can dig into the approach vueify is using

**Comment by Xiphe** [comment]
Hey @vire. I digged into it a bit but gave up now. It's far more complex then I thought since we have about 4 different code manipulation steps that need source mapping and only babel+ts support that natively.

 Steps I identified:
  - component parsing via `vue-template-compiler`
  - template compiling via `vue-template-compiler`
  - js transform via `babel` or `typescript`
  - custom merge of transformed JS, render functions, etc. in `generateOutput`

Additionally, I was not able to find documentation on how jest takes sourcemaps, other common preprocessors seem to not provide source maps, too.

**Comment by eddyerburgh** [comment]
To pass source maps to Jest you return an object like this from `process`:

```
module.exports = {
  process (src, path) {
  ...
    return {
      code: output,
      map: sourceMap
    }
  }
}
```

For the source map to be used by Jest, you need to set `mapCoverage` to `true` - https://facebook.github.io/jest/docs/configuration.html#mapcoverage-boolean

You can generate a sourcemap using the source-map module. As you identified, if there are existing source maps you need to combine these with the new source map you generate.

You can see this done in vueify - https://github.com/vuejs/vueify/blob/master/lib/compiler.js#L187

**Comment by Xiphe** [comment]
nice! totally missed out on the `mapCoverage` option. That explains why I didn't see any change at all :) Will dig deeper when I find the time. Thanks for the hint @eddyerburgh

**Comment by Xiphe** [comment]
I gave this another try the spike of @johnsardine works quite nice (once `mapCoverage` is enabled 😬 )

Another problem is that [`generateOutput`](https://github.com/vire/jest-vue-preprocessor/blob/master/index.js#L66-L86) is shifting the source mapping so I wrote a babel inline plugin for that.

My current state is really dirty but works for my usecase

https://github.com/sumcumo/jest-vue-preprocessor/tree/support-sourcemaps

Following things need to be done before I open a PR

 - [ ] cleanup
 - [ ] support ts
 - [ ] fix existing tests
 - [ ] add tests for new behaviour

**Comment by Xiphe** [comment]
I guess with [vue@2.4](https://vuejsdevelopers.com/2017/07/17/vue-js-2-4-important-features/) we could also stop [converting es2015 modules to common js](https://github.com/vire/jest-vue-preprocessor/blob/master/index.js#L69-L72)

**Comment by Xiphe** [comment]
Disclaimer: typescript sourcemaps are still off when I tested it. Not sure how to fix that I propose someone with more typescript know how looks into that.

**Comment by DarynHolmes** [comment]
I also have a missing component from my Jest coverage. This is similar to @gkatsanos. I am sure it was showing at one stage, I recently noticed that it is missing...

**Comment by gkatsanos** [comment]
@DarynHolmes try using this option http://facebook.github.io/jest/docs/en/configuration.html#collectcoveragefrom-array (check the webpack vue-cli template config , probably you can just copy paste it: https://github.com/vuejs-templates/webpack/blob/develop/template/test/unit/jest.conf.js )

**Comment by felipebastosweb** [comment]
in the jest.conf.js work fine

//mapCoverage: true,
  testURL: 'http://localhost'
